### PR TITLE
Adding CHPs to the existing heating baseyear

### PR DESCRIPTION
**Pull request by toniseibold**
Closes # (if applicable).

## Adding CHPs to the existing heating sector 
In `add_existing_baseyear.py`, the code now checks for the wildcard `H`. If the heating sector is not part of the network, CHP plants are added to the electricity grid only. Otherwise, the CHPs are added with the heating sector. If the column `Capacity_thermal` exists, the MaStR data was read in and the German CHPs are added with the efficiency calculated by Triebs et al.

## Checklist

- [x] I tested my contribution locally and it seems to work fine.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in all of `config.default.yaml`.
- [x] Changes in configuration options are also documented in `doc/configtables/*.csv`.
- [ ] A release note `doc/release_notes.rst` is added.

### CODE REVIEW

**Comment by fneum** [comment]
- hydro and waste are dropped later anyways, hence I dropped them from the powerplant query
- `skipna=True` in mean calculation is not required as this is already the default behaviour
- `try-except` block for lifetimes was not required in my case; if it is needed, I would implement as `if-else` block
- hardcoded heat efficiency (0.71) was removed by using the backpressure coefficient (https://ens.dk/sites/ens.dk/files/Analyser/technology_data_catalogue_for_el_and_dh.pdf, page 21: "The Cb -coefficient (backpressure coefficient) is defined as the maximum power generation capacity in backpressure mode divided by the maximum heat production capacity (including flue gas condensation if applicable)")
- read custom powerplants here, rather than in `build_powerplants` rule.
- I added a mechanism to exclude small CHPs below 500 kW (threshold debatable)
- I use the backpressure coefficient `c_b` for calculating the heat efficiency of power-driven CHPs and followed roughly your approach for strongly heat-driven CHPs while adjusting the split and giving some upper limits for `c_b` and `Efficiency` depending on `efficiency-heat`.
- It should be fine for `DateIn` and `DateOut` to be floats.
- There was a bug which only used the first capacity value in a pandas.Series (`p_nom=capacity[0],` and `capacity[0] / costs.at[key, "efficiency"]`)

TODOs

- [ ] Double-check my changes
- [ ] Handle Fueltypes Other/Waste (relatively large capacity dropped, also linked to duplicates in `pypsa-ariadne` PR)
- [ ] Potentially: keep oil CHPs and use gas CHP for techno-economic data
- [ ] Test run with capacity and energy output comparison to historical statistics.

**Comment by toniseibold** [comment]
in `add_existing_baseyear.py` why are oil CHPs kept?
```
    if "H" in snakemake.wildcards.sector_opts.split("-"):
        df_agg = df_agg.query("Set != 'CHP' or Fueltype == 'Oil'")
```
Oil and waste CHPs are now kept in the heating sector.

File: scripts/add_existing_baseyear.py [hunk 1]
```diff
@@ -809,6 +811,7 @@ def add_heating_capacities_installed_before_baseyear(
 
     baseyear = snakemake.params.baseyear
 
+    # n = pypsa.Network('/home/toni-seibold/Documents/CHP/pypsa-ariadne/results/240209-365H-fixcoalco2/prenetworks/elec_s_22_lv1.2__365H-T-H-B-I-A_2020.nc')
```
Comment (fneum) [inline scripts/add_existing_baseyear.py:814]: remove accidental addition

**Review by fneum — APPROVED** [review]
This looks good. Maybe add a release note to `doc/release_notes.rst` and remove the accidental file path added.

File: scripts/add_existing_baseyear.py [hunk 2]
```diff
@@ -142,6 +142,9 @@ def add_power_capacities_installed_before_baseyear(n, grouping_years, costs, bas
     # Assume that all oil power plants are not CHPs.
     if "H" in snakemake.wildcards.sector_opts.split("-"):
         df_agg = df_agg.query("Set != 'CHP'")
+    else:
+        if "I" not in snakemake.wildcards.sector_opts.split("-"):
+            df_agg.query("Industry" == False, inplace=True)
```
Comment (fneum) [inline scripts/add_existing_baseyear.py:147]:
```suggestion
    elif "I" not in snakemake.wildcards.sector_opts.split("-") and "Industry" in df_agg.columns:
         df_agg.query("not Industry", inplace=True)
```
very minor :)

But the code should check for existence of "Industry" column, as this cannot be assumed outside the ariadne context.

File: scripts/add_existing_baseyear.py [hunk 3]
```diff
@@ -616,6 +619,9 @@ def add_heating_capacities_installed_before_baseyear(
 
     # check if the CHPs were read in from MaStR for Germany
     if "Capacity_thermal" in chp.columns:
+        if "I" not in snakemake.wildcards.sector_opts.split("-"):
+            chp.query("Industry == False", inplace=True)
```
Comment (fneum) [inline scripts/add_existing_baseyear.py:623]:
```suggestion
            chp.query("not Industry", inplace=True)
```
this should also work.

**Review by fneum — CHANGES_REQUESTED** [review]
one minor item left, otherwise good to go

**Comment by toniseibold** [comment]
Installed capacity of CHP plants in Germany 2020:
```
urban central coal CHP                  24.096386
urban central gas CHP                 1292.031899
urban central gas CHP CC                 0.000000
urban central lignite CHP              561.611424
urban central solid biomass CHP         42.487920
urban central solid biomass CHP CC       0.000000
```

Heating supply 
```
urban central air heat pump             4.255723e-06
urban central gas CHP                   4.075471e-06
urban central gas CHP CC                1.599523e-06
urban central gas boiler                8.522301e+07
urban central resistive heater          5.342906e-06
urban central solar thermal             8.890480e-07
urban central solid biomass CHP         6.361950e+03
```

**Comment by lindnemi** [comment]
please add the CHPs in add_existing_power and not in add_existing_heating (because of the differences in the grouping year logic)

**Comment by lindnemi** [comment]
- [ ] also make sure not to add power-only biomass plants twice

**Pull request merged**